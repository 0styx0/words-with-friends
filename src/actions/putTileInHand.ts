import types from './types';
import Tile from '../interfaces/Tile';
import PlayerClass from '../classes/Player';

export default function putTileInHand(Players: PlayerClass[], tile: Tile) {
    return {
        type: types.PLACE_TILE_IN_HAND,
        tile,
        Players
    };
}