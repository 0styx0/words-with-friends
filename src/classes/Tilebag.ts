import Tile from '../interfaces/Tile';

export default class Tilebag {

    // http://www.thewordfinder.com/wwf-point-values.php
    static readonly alphabet = new Set<(Tile & { amount: number })>([
        {
            letter: 'A', // default is A even though can be changed
            points: 0,
            amount: 2
        },
        {
            letter: 'A',
            points: 1,
            amount: 9
        },
        {
            letter: 'B',
            points: 4,
            amount: 2
        },
        {
            letter: 'C',
            points: 4,
            amount: 2
        },
        {
            letter: 'D',
            points: 2,
            amount: 5
        },
        {
            letter: 'E',
            points: 1,
            amount: 13
        },
        {
            letter: 'F',
            points: 4,
            amount: 2
        },
        {
            letter: 'G',
            points: 3,
            amount: 3
        },
        {
            letter: 'H',
            points: 3,
            amount: 4
        },
        {
            letter: 'I',
            points: 1,
            amount: 8
        },
        {
            letter: 'J',
            points: 10,
            amount: 1
        },
        {
            letter: 'K',
            points: 5,
            amount: 1
        },
        {
            letter: 'L',
            points: 2,
            amount: 4
        },
        {
            letter: 'M',
            points: 4,
            amount: 2
        },
        {
            letter: 'N',
            points: 2,
            amount: 5
        },
        {
            letter: 'O',
            points: 1,
            amount: 8
        },
        {
            letter: 'P',
            points: 4,
            amount: 2
        },
        {
            letter: 'Q',
            points: 10,
            amount: 1
        },
        {
            letter: 'R',
            points: 1,
            amount: 6
        },
        {
            letter: 'S',
            points: 1,
            amount: 5
        },
        {
            letter: 'T',
            points: 1,
            amount: 7
        },
        {
            letter: 'U',
            points: 2,
            amount: 4
        },
        {
            letter: 'V',
            points: 5,
            amount: 2
        },
        {
            letter: 'W',
            points: 4,
            amount: 2
        },
        {
            letter: 'X',
            points: 8,
            amount: 1
        },
        {
            letter: 'Y',
            points: 3,
            amount: 2
        },
        {
            letter: 'Z',
            points: 10,
            amount: 1
        }
    ]);

    tiles: Tile[] = [];

    /**
     * Generates all tiles needed in Tilebag
     */
    constructor() {

        this.tiles = Array.from(Tilebag.alphabet).reduce((tiles, letter) => {

            for (let i = 0; i < letter.amount; i++) {
                tiles = tiles.concat(letter);
            }

            return tiles;

        }, [] as Tile[]);
    }

    /**
     * Gets a tile and removes it from the Tilebag
     */
    getRandomTile(playerIndex: number) {

        const tileToFind = this.tiles[Math.floor(Math.random() * this.tiles.length)];

        const tileToGet = this.removeTile(tileToFind);

        tileToGet.playerIndex = playerIndex;
        return tileToGet;
    }

    /**
     * Removes tile from bag
     */
    private removeTile(tile: Tile): Tile {

        const positionOfTile = this.tiles.indexOf(tile);
        return JSON.parse(JSON.stringify(this.tiles.splice(positionOfTile, 1)[0]));
    }
}
