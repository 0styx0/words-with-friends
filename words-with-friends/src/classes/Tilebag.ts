import Tile from '../interfaces/Tile';

export default class Tilebag {

    public static tiles: Tile[] = [];

    // http://www.thewordfinder.com/wwf-point-values.php
    private static readonly alphabet = new Set<(Tile & { amount: number })>([
        {
            letter: '',
            points: 0,
            amount: 2
        },
        {
            letter: 'A',
            points: 1,
            amount: 9
        },
        {
            letter: 'B',
            points: 4,
            amount: 2
        },
        {
            letter: 'C',
            points: 4,
            amount: 2
        },
        {
            letter: 'D',
            points: 2,
            amount: 5
        },
        {
            letter: 'E',
            points: 1,
            amount: 13
        },
        {
            letter: 'F',
            points: 4,
            amount: 2
        },
        {
            letter: 'G',
            points: 3,
            amount: 3
        },
        {
            letter: 'H',
            points: 3,
            amount: 4
        },
        {
            letter: 'I',
            points: 1,
            amount: 8
        },
        {
            letter: 'J',
            points: 10,
            amount: 1
        },
        {
            letter: 'K',
            points: 5,
            amount: 1
        },
        {
            letter: 'L',
            points: 2,
            amount: 4
        },
        {
            letter: 'M',
            points: 4,
            amount: 2
        },
        {
            letter: 'N',
            points: 2,
            amount: 5
        },
        {
            letter: 'O',
            points: 1,
            amount: 8
        },
        {
            letter: 'P',
            points: 4,
            amount: 2
        },
        {
            letter: 'Q',
            points: 10,
            amount: 1
        },
        {
            letter: 'R',
            points: 1,
            amount: 6
        },
        {
            letter: 'S',
            points: 1,
            amount: 5
        },
        {
            letter: 'T',
            points: 1,
            amount: 7
        },
        {
            letter: 'U',
            points: 2,
            amount: 4
        },
        {
            letter: 'V',
            points: 5,
            amount: 2
        },
        {
            letter: 'W',
            points: 4,
            amount: 2
        },
        {
            letter: 'X',
            points: 8,
            amount: 1
        },
        {
            letter: 'Y',
            points: 3,
            amount: 2
        },
        {
            letter: 'Z',
            points: 10,
            amount: 1
        }
    ]);

    /**
     * Generates all tiles needed in Tilebag
     */
    static init() {

        Tilebag.tiles = Array.from(Tilebag.alphabet).reduce((tiles, letter) => {

            for (let i = 0; i < letter.amount; i++) {

                tiles = tiles.concat(letter);
            }

            return tiles;

        }, [] as Tile[]);
    }

    /**
     * Removes tile from bag
     */
    static removeTile(tile: Tile) {

        const positionOfTile = Tilebag.tiles.indexOf(tile);
        Tilebag.tiles.splice(positionOfTile - 1, 1);
    }

    /**
     * Gets a tile and removes it from the Tilebag
     */
    static getRandomTile() {

        const tileToGet = this.tiles[Math.floor(Math.random() * this.tiles.length)];
        Tilebag.removeTile(tileToGet);

        return tileToGet;
    }

}
