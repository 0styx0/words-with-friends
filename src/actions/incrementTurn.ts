import { Turn } from './interfaces';
import types from './types';
import Tilebag from '../classes/Tilebag';

export default function incrementTurn(currentTurn: number, tilebag: Tilebag): Turn {
    return {
        type: types.INCREMENT_TURN,
        turn: currentTurn,
        Tilebag: tilebag
    };
}