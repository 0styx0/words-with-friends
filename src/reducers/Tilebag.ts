import { ResetTilebag } from '../actions/interfaces';
import types from '../actions/types';
import TilebagClass from '../classes/Tilebag';

export default function Tilebag(tilebag: any = {}, action: ResetTilebag) {

    switch (action.type) {

        case types.RESET_TILEBAG:
            return new TilebagClass();
        default:
            return tilebag;
    }
}