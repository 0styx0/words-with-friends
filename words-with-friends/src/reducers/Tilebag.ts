import { Turn } from '../actions/interfaces';

export default function Tilebag(tilebag: any = {}, action: Turn) {

    return tilebag;
}