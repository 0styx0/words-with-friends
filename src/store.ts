import reducers from './reducers';
import { createStore, compose } from 'redux';
import Player from './classes/Player';
import Tilebag from './classes/Tilebag';
import Board from './classes/Board';

const enhancers = compose('devToolsExtension' in window ?
    (window as {} as { devToolsExtension: Function }).devToolsExtension!() :
    (f: {}) => f);

export const defaultState = {
    turn: 0,
    Players: [new Player(true, 0), new Player(false, 1)], // turns switched around on INIT_PLAYERS
    board: new Board(),
    Tilebag: new Tilebag()
};

const store = createStore(reducers, defaultState, enhancers);

export default store;
