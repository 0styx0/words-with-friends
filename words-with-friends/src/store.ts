import reducers from './reducers';
import { createStore } from 'redux';

export const defaultState = {
    turn: 0
};

const store = createStore(reducers, defaultState);

export default store;
