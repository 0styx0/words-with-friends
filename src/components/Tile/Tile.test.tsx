import * as React from 'react';
import * as renderer from 'react-test-renderer';
import * as casual from 'casual';
import * as sinon from 'sinon';
import store, { defaultState, getState } from '../../store';
import { Provider } from 'react-redux';
import ConnectedTileContainer, { TileContainer } from './Tile';
import mathMock from '../../test/mocks/Math';
import { mount } from 'enzyme';
import putTileInHand from '../../actions/putTileInHand';
import incrementTurn from '../../actions/incrementTurn';
import Tilebag from '../../classes/Tilebag';

mathMock();

describe('<TileContainer />', () => {

    describe('renders correctly for', () => {

        function snap(playerIndex: number) {

            const state = getState() as typeof defaultState;

            const tree = renderer.create(
                <Provider store={store}>
                    <ConnectedTileContainer
                        coordinates={[casual.integer, casual.integer]}
                        tile={(new Tilebag()).getRandomTile(playerIndex)}
                        removeTile={sinon.mock}
                        {...state as any}
                    />
                </Provider>
            ).toJSON();

            expect(tree).toMatchSnapshot();
        }

        it('current player', () => snap(0));

        it('non-current player', () => snap(1));
    });

    function setup(
        removeTile: Function = sinon.spy(),
        coordinates: ReadonlyArray<number> | undefined = [casual.integer(), casual.integer()]
    ) {

        const state = store.getState() as typeof defaultState;

        const wrapper = mount(
            <Provider store={store}>
                <ConnectedTileContainer
                    coordinates={coordinates}
                    tile={state.Tilebag.getRandomTile(0)}
                    {...state as any}
                    removeTile={removeTile}
                />
            </Provider>
        );

        return {
            wrapper,
            removeTile: removeTile as sinon.SinonSpy
        };
    }

    function drag() {

        const { wrapper } = setup();

        const dragEventMocks = {
            dataTransfer: {
                setData: sinon.mock(),
                dropEffect: ''
            }
        };

        wrapper.find(TileContainer).simulate('dragstart', dragEventMocks);

        return {
            wrapper,
            dragEventMocks,
            component: wrapper.find(TileContainer).instance()
        };
    }

    describe('#onDragStart', () => {

        it(`changes dropEffect to 'move' and sets data to stringified tile`, () => {

            const { dragEventMocks, component } = drag();

            expect(
                dragEventMocks
                    .dataTransfer
                    .setData.calledWith('tile', JSON.stringify(component.props.tile))
            ).toBeTruthy();

            expect(dragEventMocks.dataTransfer.dropEffect).toBe('move');
        });
    });

    describe('#onDragEnd', () => {

        it(`removes tile if dropEffect is 'move'`, () => {

            const { wrapper, removeTile } = setup();

            wrapper.find(TileContainer).simulate('dragend', { dataTransfer: { dropEffect: 'move' } });

            expect(removeTile.called).toBeTruthy();
        });

        it(`doesn't remove tile if dropEffect isn't 'move'`, () => {

            const { wrapper, removeTile } = setup();

            wrapper.find(TileContainer).simulate('dragend', { dataTransfer: { dropEffect: 'none' } });

            expect(removeTile.called).toBeFalsy();
        });
    });

    describe('#canDrag', () => {

        describe('returns true if', () => {

            it(`tile in current player's hand`, () => {

                const { wrapper } = setup(sinon.mock(), undefined);

                const component = wrapper.find(TileContainer).instance() as TileContainer;

                expect(component.canDrag()).toBeTruthy();
            });

            it('tile is on board but placed during current turn', () => {

                const coordinates: ReadonlyArray<number> = [casual.integer(), casual.integer()];

                const { wrapper } = setup(sinon.mock(), coordinates);

                const component = wrapper.find(TileContainer).instance() as TileContainer;

                expect(component.canDrag()).toBeTruthy();
            });
        });

        describe('return false if', () => {

            it(`tile in non-current player's hand`, () => {

                const { wrapper } = setup(sinon.mock(), undefined);
                const state = store.getState() as typeof defaultState;

                const component = wrapper.find(TileContainer).instance() as TileContainer;

                store.dispatch(putTileInHand(state.Players[1], component.props.tile));

                expect(component.canDrag()).toBeFalsy();
            });

            it('tile is on board and placed there during different turn', () => {

                const state = store.getState() as typeof defaultState;
                const coordinates: ReadonlyArray<number> = [casual.integer(), casual.integer()];

                const { wrapper } = setup(sinon.mock(), coordinates);

                store.dispatch(incrementTurn(state.turn, state.Tilebag));

                const component = wrapper.find(TileContainer).instance() as TileContainer;

                expect(component.canDrag()).toBeFalsy();
            });
        });
    });
});