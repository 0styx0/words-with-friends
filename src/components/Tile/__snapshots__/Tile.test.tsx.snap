// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TileContainer /> renders correctly for current player 1`] = `
<div
  className="tileContainer"
  draggable={true}
  onDragEnd={[Function]}
  onDragStart={[Function]}
>
  <div
    className="tile"
  >
    <span
      className="points"
    >
      2
    </span>
    <span
      className="letter"
    >
      L
    </span>
  </div>
</div>
`;

exports[`<TileContainer /> renders correctly for non-current player 1`] = `
<div
  className="tileContainer"
  draggable={false}
  onDragEnd={[Function]}
  onDragStart={[Function]}
>
  <div
    className="tile"
  >
    <span
      className="points"
    >
      5
    </span>
    <span
      className="letter"
    >
      K
    </span>
  </div>
</div>
`;
